File: src/main/java/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/producer/ArgUnpackerProducer.java
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.producer;

import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.unpacker.UnpackerParser;
import io.github.oleksivio.telegram.bot.core.controller.handler.unpack.UpdateUnpacker;
import io.github.oleksivio.telegram.bot.core.model.ITelegram;
import org.springframework.stereotype.Controller;

import java.lang.reflect.Method;
import java.util.Collection;
import java.util.Optional;

@Controller
public class ArgUnpackerProducer {

    private final Collection<UnpackerParser> unpackerParsers;

    public ArgUnpackerProducer(Collection<UnpackerParser> unpackerParsers) {
        this.unpackerParsers = unpackerParsers;
    }

    public <T extends ITelegram> Optional<UpdateUnpacker<T>> create(Method method) {
        for (UnpackerParser unpackerParser : unpackerParsers) {
            if (unpackerParser.canParse(method)) {

                UpdateUnpacker<T> resultUnpacker = unpackerParser.getUpdateUnpacker();

                return Optional.of(resultUnpacker);
            }
        }

        return Optional.empty();
    }
}

File: src/main/kotlin/io/github/oleksivio/telegram/bot/core/controller/annotations/parser/producer/ArgUnpackerProducer.kt
package io.github.oleksivio.telegram.bot.core.controller.annotations.parser.producer

import io.github.oleksivio.telegram.bot.core.controller.annotations.parser.unpacker.UnpackerParser
import io.github.oleksivio.tl.kbot.server.api.model.ITelegram
import org.springframework.stereotype.Controller
import kotlin.reflect.KFunction

@Controller
class ArgUnpackerProducer(private val unpackerParsers: Collection<UnpackerParser<*, *>>) {

    fun <T : ITelegram> create(func: KFunction<*>): UnpackerParser<Annotation, T>? {
        for (unpackerParserAbstract in unpackerParsers) {
            if (unpackerParserAbstract.canParse(func)) {
                return unpackerParserAbstract as UnpackerParser<Annotation, T>
            }
        }

        return null
    }
}
